### PR TITLE
SA1128ConstructorInitializerMustBeOnOwnLine

### LINKED ISSUE #1177: New rule proposal: Constructor initializer must appear on its own line
**Category:** Readability
**Title:** Constructor initializer must appear on its own line
**Description:** A violation of this rule occurs when an instance constructor contains a constructor initializer, but the constructor initializer is not on its own line. For example:

```csharp
public class TypeName
{
    public TypeName() : this(0)
    {
    }

    public TypeName(int value)
    {
    }
}
```

To resolve a violation of this rule, move the constructor initializer to its own line, including the `:` character.

```csharp
public class TypeName
{
    public TypeName()
        : this(0)
    {
    }

    public TypeName(int value)
    {
    }
}
```

:thought_balloon: I'm thinking that violations of this rule could be specifically determined by the placement of the `:`, meaning the following would violate this as well. However, I'd be open to alternative views if people think this would be a major problem.

```csharp
public class TypeName
{
    public TypeName() :
        this(0)
    {
    }

    public TypeName(int value)
    {
    }
}
```

### PR DESCRIPTION
**Pull request by nbarbettini**
Fixes #1177

### CODE REVIEW

**Comment by sharwell** [comment]
Blocked on #1334 (for diagnostic numbering).

File: documentation/SA1128.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+{
+    public TypeName() : this(0)
+    {
+    }
+
+    public TypeName(int value)
+    {
+    }
+}
+#pragma warning restore SA1128 // Constructor initializers must be on their own line
```
Comment (sharwell) [inline documentation/SA1128.md]: :bulb: Try to restrict your suppression to the necessary line(s).

```csharp
public class TypeName
{
#pragma warning disable SA1128 // Constructor initializers must be on their own line
    public TypeName() : this(0)
#pragma warning restore SA1128 // Constructor initializers must be on their own line
    {
    }

    public TypeName(int value)
    {
    }
}
```

Comment (nbarbettini) [inline documentation/SA1128.md]: Got it, makes sense. Thanks for all the comments, it's a big help.

**Comment by sharwell** [comment]
:question: @nbarbettini Did you want to finish the code fix before merging this, or leave it as an enhancement request?

**Comment by nbarbettini** [comment]
Leave as an enhancement request, for now.

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1128ConstructorInitializerMustBeOnOwnLine.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    using System.Collections.Immutable;
+    using System.Linq;
+    using Helpers;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Diagnostics;
+
+    /// <summary>
+    /// Todo
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1128ConstructorInitializerMustBeOnOwnLine.cs]: :bulb: Missing comment.

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1128ConstructorInitializerMustBeOnOwnLine.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            var initializer = constructor.Initializer;
+            var colon = initializer.ColonToken;
+
+            if (!colon.IsFirstInLine())
+            {
+                context.ReportDiagnostic(Diagnostic.Create(Descriptor, initializer.GetLocation()));
+                return;
+            }
+
+            if (colon.TrailingTrivia.Any(t => t.IsKind(SyntaxKind.EndOfLineTrivia)))
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1128ConstructorInitializerMustBeOnOwnLine.cs]: :bulb: `.Any(SyntaxKind.EndOfLineTrivia)` is a more efficient overload (fixed during merge).

File: documentation/SA1128.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+}
+```
+
+The following code would not produce any violations:
+
+```csharp
+public class TypeName
+{
+    public TypeName() :
+        this(0)
```
Comment (sharwell) [inline documentation/SA1128.md]: :bulb: The sample still breaks the rule (fixed during the merge).

File: documentation/SA1128.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+```
+
+## How to fix violations
+
+To fix a violation of this rule, the constructor initializer, including the `:` character, to its own line.
+
+## How to suppress violations
+
+```csharp
+#pragma warning disable SA1128 // Constructor initializers must be on their own line
```
Comment (sharwell) [inline documentation/SA1128.md]: :bulb: These directives can be placed around just the line containing the `:` (fixed during the merge).

**Comment by nbarbettini** [comment]
D'oh, thanks for fixing those things I overlooked :+1:

**Pull request merged**